feat: convert to typescript

achingbrain opened the pull request:
Converts this module to typescript.  Still a few things to do:

- [ ] Fill out this PR description with list of changes
- [ ] Examples passing
- [ ] Interop passing

BREAKING CHANGE: types are no longer hand crafted, this module is now ESM only

BigLep commented:
2022-03-22 conversation for someone reviewing this code:
1. Check it out 
2. See if can integrate it with another project.

mpetrunic commented on .github/workflows/main.yml:
```diff
@@ -5,40 +5,23 @@ on:
   pull_request:
     branches:
-      - '**'
+      - master
```
isn't ci suppose to run even if you open PR against some old branch. For example if backporting fix to older version?

achingbrain replied:
Yes, I'll revert this change.  Probably worth opening a PR to fix this in the automated CI flows too, [the line is here](https://github.com/protocol/.github/blob/master/templates/.github/workflows/js-test-and-release.yml#L8).

mpetrunic commented on package.json:
```diff
@@ -2,126 +2,228 @@
       "*.d.ts"
     ]
   },
+  "release": {
```
Seems like libp2p repo doesn't use semantic release so this is useless

achingbrain replied:
Good spot - it was added by [check-aegir-project](https://www.npmjs.com/package/check-aegir-project), I've updated it to only do that if there's an npm script called `"release"` with the contents `"semantic-release"`

mpetrunic commented on package.json:
```diff
@@ -2,126 +2,228 @@
+    "test:firefox-webworker": "npm run test -- -t webworker -f \"./dist/test/**/*.spec.js\" -- --browser firefox",
+    "test:examples": "cd examples && npm run test:all",
+    "test:interop": "npm run test -- -t node -f dist/test/interop.js",
+    "release": "semantic-release"
```
don't think we have dependency and it seems like its not used in CD

achingbrain replied:
The dep is added by aegir but you are quite right it's not used.

mpetrunic commented on src/circuit/auto-relay.ts:
```diff
@@ -0,0 +1,284 @@
+   * If the protocol is supported, check if the peer supports **HOP** and add it as a listener if
+   * inside the threshold.
+   */
+  async _onProtocolChange (evt: CustomEvent<PeerProtocolsChangeData>) {
```
We don't use private because of tests or is it too much effort to change them to private?

achingbrain replied:
The external API for the components are largely defined by interfaces, but it's useful to be able to call internal methods in the tests, but yes, this could change to be private.

I tried out [hash-prefixed private](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields) in the [EventEmitter](https://github.com/libp2p/js-libp2p-interfaces/blob/master/packages/libp2p-interfaces/src/index.ts#L77) class - it works but causes a warning to be logged to the console during the tests, kind of want to experiment with that a bit more before rolling it out everywhere.

In the interim my hope is the interface-as-public api should stop people calling private(ish) methods.

mpetrunic commented on src/config.ts:
```diff
@@ -0,0 +1,101 @@
+import errCode from 'err-code'
+import type { RecursivePartial } from '@libp2p/interfaces'
+
+const DefaultConfig: Partial<Libp2pInit> = {
```
How come you put Partial here? 

Shouldn't we enforce that whoever adds a new module must put defaults here? Maybe even create some type which would make everything mandatory here 😄

achingbrain replied:
Without `Partial` is becomes necessary to define things things like `.peerId` and `.datastore` that aren't simple values.

mpetrunic replied:
```suggestion
const DefaultConfig: Required<Omit<Libp2pInit, "peer" | "datastore">> & Libp2pInit = {
```
would probably do the trick but I don't have a strong preference, it might end up being too hard to maintain what to keep optional.

achingbrain replied:
Yeah, it starts getting a bit fiddly like this, maybe let's look at it again later.

mpetrunic commented on src/errors.ts:
```diff
@@ -1,17 +1,19 @@
 }
 
-exports.codes = {
+export const codes = {
```
enum?

achingbrain replied:
Enums end up as numbers at runtime, right?  the use case is:

```js
throw errCode(new Error(messages.FOO), codes.FOO)

// later
if (err.code === codes.FOO) {
  ...
}
```

If someone decided to re-order the enum (alphabetical, add a new code, whatever) this could cause very undesirable side effects in code that did (rightly or wrongly):

```js
if (err.code === 'FOO') {
  ...
}
```

achingbrain replied:
Non-ts code, I mean.  Maybe I'm worrying about nothing.

mpetrunic replied:
Hmm we can use string enums so they don't end up being numbers and have same functionality as we have now: https://www.typescriptlang.org/docs/handbook/enums.html#string-enums

And they shouldn't be replaced by value unless you add const: https://www.typescriptlang.org/docs/handbook/enums.html#const-enums

achingbrain replied:
Ah, nice - I've updated it.

wemeetagain commented on doc/CONFIGURATION.md:
```diff
@@ -235,189 +235,177 @@ Besides the `modules` and `config`, libp2p allows other internal options and con
+
+const node = await createLibp2p({
+  transports: [
+    TCP,
```
I think this doesn't work anymore, same below where classes are used instead of instances.

wemeetagain commented on doc/CONFIGURATION.md:
```diff
@@ -719,19 +693,17 @@ const node = await Libp2p.create({
+import { Mplex } from '@libp2p/mplex'
+import { Noise } from '@chainsafe/libp2p-noise'
+
+const { FaultTolerance } from 'libp2p/src/transport-manager')
```
invalid syntax

wemeetagain commented on doc/STREAMING_ITERABLES.md:
````diff
@@ -22,8 +22,8 @@
 ```js
-const duplexPair = require('it-pair/duplex')
-const pipe = require('it-pipe')
+const duplexPair from 'it-pair/duplex')
````
syntax error

wemeetagain commented on doc/migrations/v0.26-v0.27.md:
````diff
@@ -80,7 +82,7 @@ libp2p.dialProtocol(peerInfo, '/echo/1.0.0', (err, conn) => {
 **After**
 ```js
-const pipe = require('it-pipe')
+const pipe from 'it-pipe')
````
syntax

wemeetagain reviewed (commented):
Update LGTM. Some small comments regarding code snippets in docs. We should consider adding something like https://github.com/bbc/typescript-docs-verifier to the CI pipeline.

achingbrain commented:
@wemeetagain thanks for your review - I've addressed the points in #1182

[bbc/typescript-docs-verifier](https://github.com/bbc/typescript-docs-verifier) - this looks really useful, I didn't know it existed.  Would you like to give adding it to CI a go?

wemeetagain commented:
Ok, I'll open a PR
